src/skadi/sample.cc

```diff
@@ -431,74 +419,135 @@ template <class coords_t> std::vector<double> sample::get_all(const coords_t& co
 
     if (index != curIndex) {
+      std::lock_guard<std::mutex> _(cache_lck);
       tile = cache_->source(index);
       curIndex = index;
     }
 
-    double value = NO_DATA_VALUE;
+    double value{get_no_data_value()};
     if (tile) {
       double u = (coord.first - lon) * (HGT_DIM - 1);
       double v = (1.0 - (coord.second - lat)) * (HGT_DIM - 1);
       value = tile.get(u, v);
     }
 
+    if (value == get_no_data_value())
+      value = get_from_remote(coord);
+
     values.emplace_back(value);
   }
+
   return values;
 }
 
-double sample::get_no_data_value() {
-  return NO_DATA_VALUE;
-}
+template <class coord_t> double sample::get_from_cache(const coord_t& coord) {
+  std::lock_guard<std::mutex> _(cache_lck);
+  auto tile = cache_->source(get_tile_index(coord));
+  if (!tile)
+    return get_no_data_value();
 
-template <class coord_t> uint16_t sample::get_tile_index(const coord_t& coord) {
-  auto lon = std::floor(coord.first);
-  auto lat = std::floor(coord.second);
-  return static_cast<uint16_t>(lat + 90) * 360 + static_cast<uint16_t>(lon + 180);
+  double u = (coord.first - std::floor(coord.first)) * (HGT_DIM - 1);
+  double v = (1.0 - (coord.second - std::floor(coord.second))) * (HGT_DIM - 1);
+  return tile.get(u, v);
 }
 
-void sample::add_single_tile(const std::string& path) {
-  cache_->cache.front().init(path, format_t::RAW);
+bool sample::store(const std::string& elev, const std::vector<char>& raw_data) {
+  // data_source never changes so we do not lock it. it is set only in sample constructor
+  auto fpath = cache_->data_source + elev;
+  if (filesystem::exists(fpath))
+    return true;
+
+  auto data = cache_item_t::parse_hgt_name(elev);
+  if (!data || data->second == format_t::UNKNOWN)
+    return false;
+
+  // cache size is always the same.
+  if (data->first >= cache_->size())
+    return false;
+
+  if (!filesystem::save(fpath, raw_data))
+    return false;
+
+  std::lock_guard<std::mutex> _(cache_lck);
+  return cache_->insert(data->first, fpath, data->second);
 }
 
-std::string sample::get_hgt_file_name(uint16_t index) {
-  auto x = (index % 360) - 180;
-  auto y = (index / 360) - 90;
+template <class coord_t> double sample::get_from_remote(const coord_t& coord) {
+  if (url_.empty() || !remote_loader_)
+    return get_no_data_value();
 
-  std::string name(y < 0 ? "/S" : "/N");
-  y = std::abs(y);
-  if (y < 10) {
-    name.push_back('0');
+  auto index = get_tile_index(coord);
+  auto elev = get_hgt_file_name(index);
+  auto uri = baldr::make_single_point_url(url_, elev, remote_path_);
+
+  {
+    std::shared_lock<std::shared_timed_mutex> _(st_lck);
+    if (st_.count(uri))
+      return get_from_cache(coord);
   }
-  name.append(std::to_string(y));
-  name.append(name);
 
-  name.append(x < 0 ? "W" : "E");
-  x = std::abs(x);
-  if (x < 100) {
-    name.push_back('0');
+  LOG_INFO("Start loading data from remote server address: " + uri);
+  auto result = remote_loader_->get(uri);
```

Comment:
I believe there's still possible to fetch same tile from multiple threads because you split mutex-ed logic into 2 blocks
```
1) check with lock
2) remote_loader->fetch
3) mark tile as fetch with lock
```

It is possible that multiple threads pass 1) and start fetching tile in 2) at the same time. However, it's better to block all other threads if one of them has already started fetching the tile(or had intention and grabbed the lock)